test/e2e/knative_serving_test.go

```diff
@@ -0,0 +1,163 @@
+		}
+	})
+
+	t.Run("user permissions", func(t *testing.T) {
+		testUserPermissions(t, contexts)
+	})
+
+	t.Run("undeploy serverless operator and check dependent operators removed", func(t *testing.T) {
+		adminCtx.Cleanup()
+		err := test.WaitForOperatorDepsDeleted(adminCtx)
+		if err != nil {
+			t.Fatalf("Operators still running: %v", err)
+		}
+	})
+}
+
+func testUserPermissions(t *testing.T, contexts []*test.Context) {
+	editCtx := contexts[1]
+	viewCtx := contexts[2]
+	tests := []struct {
+		name          string
+		userRole      string
+		userContext   *test.Context
+		operationName string
+		operation     func(context *test.Context) error
+		wantErr       bool
+		wantErrStr    string
+	}{{
+		name:          "user with view role can get",
+		userRole:      "view",
+		operationName: "get",
+		operation: func(c *test.Context) error {
+			_, err := test.GetService(c, helloworldService, testNamespace)
+			return err
+		},
+		userContext: viewCtx,
+		wantErr:     false,
+	}, {
+		name:          "user with view role can list",
+		userRole:      "view",
+		operationName: "list",
+		operation: func(c *test.Context) error {
+			_, err := test.ListServices(c, testNamespace)
+			return err
+		},
+		userContext: viewCtx,
+		wantErr:     false,
+	}, {
+		name:          "user with view role cannot create",
+		userRole:      "view",
+		operationName: "create",
+		operation: func(c *test.Context) error {
+			_, err := test.CreateService(c, "userview-service", testNamespace, image)
+			return err
+		},
+		userContext: viewCtx,
+		wantErr:     true,
+		wantErrStr:  "is forbidden",
+	}, {
+		name:          "user with view role cannot delete",
+		userRole:      "view",
+		operationName: "delete",
+		operation: func(c *test.Context) error {
+			return test.DeleteService(c, helloworldService, testNamespace)
+		},
+		userContext: viewCtx,
+		wantErr:     true,
+		wantErrStr:  "is forbidden",
+	}, {
+		name:          "user with edit role can get",
+		userRole:      "edit",
+		operationName: "get",
+		operation: func(c *test.Context) error {
+			_, err := test.GetService(c, helloworldService, testNamespace)
+			return err
+		},
+		userContext: editCtx,
+		wantErr:     false,
+	}, {
+		name:          "user with edit role can list",
+		userRole:      "edit",
+		operationName: "list",
+		operation: func(c *test.Context) error {
+			_, err := test.ListServices(c, testNamespace)
+			return err
+		},
+		userContext: editCtx,
+		wantErr:     false,
+	}, {
+		name:          "user with edit role can create",
+		userRole:      "edit",
+		operationName: "create",
+		operation: func(c *test.Context) error {
+			_, err := test.CreateService(c, "useredit-service", testNamespace, image)
+			return err
+		},
+		userContext: editCtx,
+		wantErr:     false,
+	}, {
+		name:          "user with edit role can delete",
+		userRole:      "edit",
+		operationName: "delete",
+		operation: func(c *test.Context) error {
+			return test.DeleteService(c, "useredit-service", testNamespace)
```

Comment:
What sets this service up to be deleted? Does this rely on the other test before it to pass as well?

Reply:
Yes.